src/background/services/secrets/AddressPublicKey.ts

```diff
@@ -0,0 +1,213 @@
+      secrets.secretType === SecretType.LedgerLive
+    ) {
+      assertDerivationPath(derivationPath);
+
+      const pubKeyJson = getPublicKeyFor(secrets, derivationPath, curve);
+      assertPresent(pubKeyJson, SecretsError.PublicKeyNotFound);
+
+      return AddressPublicKey.fromJSON(pubKeyJson);
+    }
+
+    if (secrets.secretType === SecretType.PrivateKey) {
+      return AddressPublicKey.fromPrivateKey(secrets.secret, curve);
+    }
+
+    throw ethErrors.rpc.internal({
+      data: {
+        reason: SecretsError.UnsupportedSecretType,
+        context: secrets.secretType,
+      },
+    });
+  }
+
+  static fromExtendedPublicKeys(
+    xpubs: ExtendedPublicKey[],
+    curve: Curve,
+    derivationPath: string,
+  ): AddressPublicKey<true> {
+    assertDerivationPath(derivationPath);
+
+    if (curve !== 'secp256k1') {
+      throw ethErrors.rpc.internal({
+        data: {
+          reason: SecretsError.UnsupportedCurve,
+          context: `"${curve}" is not supported with extended public keys`,
+        },
+      });
+    }
+
+    const matchingXpub = getExtendedPublicKeyFor(xpubs, derivationPath, curve);
+
+    if (!matchingXpub) {
+      throw ethErrors.rpc.internal({
+        data: {
+          reason: SecretsError.MissingExtendedPublicKey,
+          context: `${derivationPath} / ${curve}`,
+        },
+      });
+    }
+
+    const pathSuffix = derivationPath.slice(
+      matchingXpub.derivationPath.length + 1, // Add one to account for the trailing slash from the lookup
+    );
+    const node = fromBase58(matchingXpub.key).derivePath(pathSuffix);
+    const key = hex.encode(node.publicKey);
+
+    return new AddressPublicKey(
+      key,
+      curve,
+      derivationPath,
+      matchingXpub.btcWalletPolicyDetails,
+    );
+  }
+
+  static fromPrivateKey(
+    privateKey: string,
+    curve: Curve,
+  ): AddressPublicKey<false> {
+    let key: string;
+
+    switch (curve) {
+      case 'ed25519':
+        key = hex.encode(ed25519.getPublicKey(privateKey));
+        break;
+
+      case 'secp256k1':
+        key = hex.encode(getPublicKeyFromPrivateKey(privateKey));
+        break;
+
+      default:
+        throw ethErrors.rpc.internal({
+          data: {
+            reason: SecretsError.UnsupportedCurve,
+            context: curve,
+          },
+        });
+    }
+
+    return new AddressPublicKey(key, curve, null);
+  }
+
+  static async fromSeedphrase(
+    seedphrase: string,
+    curve: Curve,
+    derivationPath: string,
+  ): Promise<AddressPublicKey> {
+    assertDerivationPath(derivationPath);
+    const seed = await mnemonicToSeed(seedphrase);
+    let key: string;
+
+    switch (curve) {
+      case 'secp256k1': {
+        const seedNode = fromSeed(seed);
+        key = hex.encode(seedNode.derivePath(derivationPath).publicKey);
+        break;
+      }
+
+      case 'ed25519': {
+        const hdKey = slip10.fromMasterSeed(seed);
+        key = hex.encode(hdKey.derive(derivationPath).publicKey);
```

Comment:
Use `.publicKeyRaw` here -- `publicKey` is 0-prefixed (33 bytes total), we need the raw 32 bytes.